presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementFilteredAggregations.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+
+/**
+ * Implements filtered aggregations by transforming plans of the following shape:
+ *
+ * <pre>
+ * - Aggregation
+ *        F1(...) FILTER (WHERE C1(...)),
+ *        F2(...) FILTER (WHERE C2(...))
+ *     - X
+ * </pre>
+ *
+ * into
+ *
+ * <pre>
+ * - Aggregation
+ *        F1(...) mask ($0)
+ *        F2(...) mask ($1)
+ *     - Project
+ *            &lt;identity projections for existing fields&gt;
+ *            $0 = C1(...)
+ *            $1 = C2(...)
+ *         - X
+ * </pre>
+ */
+public class ImplementFilteredAggregations
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
```

Comment:
I think there will be a long time before legacy optimizers got deleted. The problem I see is that once a rule is migrated then only legacy optimizers will get updated (bugfixes, new features). 

Generally code here is copy pasted from the other rule. I wonder if it possible to somehow share the code for time being so new rules got updates also. This would probably require change the legacy rule to somehow match new optimizer interface:
```
@Override
        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Optional<Symbol>> context)
        {
            Optional<PlanNode> transformed = ...;
            if (! transformed.isPresent()) {
                return context.defaultRewrite(node, Optional.empty());
```

so the code under `...` could be somehow used here.

What do you think?